Terminal: command execution sandbox

brechtcs opened the pull request:
As discussed in https://github.com/dfurl/dterm/issues/12, this PR will port my work on dterm's execution environment to the Beaker sidebar terminal. I'm opening the PR early, so that you can already have a look, but there's still some work to do before it can be merged:

- [x] Implement the missing methods on the `DatArchive` shim
- [x] Support fully-qualified script URLs
- [x] ~~Resolve saved scripts from user directory~~ (Out of scope.)
- [x] Add more meaningful error message when command can't be resolved

brechtcs committed:
Basic implementation of command execution sandbox

brechtcs committed:
Fix string output for builtin commands

pfrazee commented on js/views/terminal.js:
```diff
@@ -198,19 +196,56 @@ class WebTerm extends LitElement {
   }
 
+  async evalBuiltin (cmd, args) {
+    let output = await this.env[cmd](this.env, ...args)
```
Oh right, I guess we don't have to `eval()` anymore.

brechtcs replied:
Yeah, I removed the `eval` from dterm a while ago and haven't had any issues.

pfrazee commented on js/lib/iframe-dom-transport.js:
```diff
@@ -0,0 +1,49 @@
+const FORBIDDEN_NODES = ['iframe', 'link', 'meta', 'script', 'style']
```
Correct me if I'm wrong but I think this should probably include

 - body
 - head
 - html
 - base
 - embed
 - object
 - canvas

More on this in a general comment

brechtcs replied:
Good point, I'll add those as well as the form controls.

pfrazee commented:
This is looking great! I'm excited to get this in. I like the approach you're using with the HTML sanitation.

We should figure out the interaction & security model of responses. Do we intend to allow interactivity in the output of commands? I think, if so, the output will need to be embedded in an iframe because it'll need to run scripts. If no, we probably shouldn't allow `<form>` or any kind of interactive element in the response html (buttons, inputs) because they won't really be able to do anything useful.

EDIT: rephrased for clarity.

pfrazee commented:
> Resolve saved scripts from user directory (Although maybe out-of-scope for this PR?)

Yeah let's save that for a later PR. We'll need the full user-filesystem model, which hasn't landed yet because it relies on Dat 2.0.

brechtcs commented:
> Do we intend to allow interactivity in the output of commands? I think, if so, the output will need to be embedded in an iframe because it'll need to run scripts. If no, we probably shouldn't allow &lt;form&gt; or any kind of interactive element in the response html (buttons, inputs) because they won't really be able to do anything useful.

For dterm I considered using HTML-only form submissions for interactivity, handling them in a way similar to traditional server-side web applications. I haven't had time yet to actually try this in practice though, so there might be some gotchas I haven't thought of. Alternatively, the interactivity API you proposed in the original Webterm spec should definitively be doable. The implementation would be very similar to how the `DatArchive` shim works.

In the scope of this PR, I think disallowing form controls is a good idea yes. Because if we go for the latter approach to implement interactivity they should be disallowed permanently, and if we go for the former it's still safer to only allow them when the feature's actually implemented.

pfrazee commented:
A forms-based interaction model would very interesting. We'd basically need to translate the form submission into a command call. Maybe scope it so that it can only call the originating command? And then the form values get translated to switches. Eg `<input name="foo" value="bar">` translates to `--foo bar`.

I agree that we leave that out of this PR but that's a super interesting idea.

brechtcs committed:
Support running command from fully-qualified URL

brechtcs committed:
Expand list of forbidden nodes for command output

brechtcs committed:
Use text node instead of <pre> for text output

brechtcs committed:
Fix sandbox communication bugs

- empty `opts` swallowed in cross-frame messaging
- onload event listener not removed because of bind

brechtcs committed:
Add missing DatArchiveShim methods

brechtcs commented on js/lib/iframe-shim-dat-archive.js:
```diff
@@ -0,0 +1,156 @@
+  async stat (...args) {
+    let stat = await request({method: 'stat', url: this.url, args})
+    stat.isDirectory = () => stat.mode === 16877
+    stat.isFile = () => stat.mode === 33188
```
This is purely based on observation, not 100% sure this is the right way to do `isFile` and `isDirectory` checks

brechtcs commented on js/lib/iframe-shim-dat-archive.js:
```diff
@@ -0,0 +1,156 @@
+
+function request (params) {
+  return new Promise(function (resolve, reject) {
+    params.token = Math.random()
```
Does `Math.random()` suffice or do we need something more solid here?

brechtcs commented on js/views/terminal.js:
```diff
@@ -198,19 +196,58 @@ class WebTerm extends LitElement {
+  evalSandbox (cmd, args) {
+    return new Promise((resolve, reject) => {
+      // Unique token to identify sandbox response
+      var token = Math.random()
```
Does `Math.random()` suffice or do we need something more solid here?

brechtcs commented:
Alright, I implemented the bulk of the missing methods on the `DatArchive` shim. The only ones still missing are `watch` and `addEventListener`, because they need to be handled slightly differently by the messaging system.

I'm inclined to leave them out of this PR though, because I don't really see a use case for them in a terminal that doesn't support streaming. 

(Sidenote: I did implement streaming in dterm using generators, but it does create some difficulties with command sub-invocations, so I don't know if we'd want to introduce them here. If yes, it would probably become useful to shim the event-based `DatArchive` methods as well. But that's an entirely seperate conversation.)

pfrazee commented:
Okay great! I'm going to give this a look soon (sometime this week). We're also going to need a permission prompt of some kind before writes occur to a dat.

brechtcs commented:
Right, the `handleDatArchiveRequests` setup function already accepts a function to determine whether to allow the request. But I'm not using it yet, because I didn't know how to embed it into the sidebar UI. Any pointers?

brechtcs commented on js/lib/iframe-shim-dat-archive.js:
```diff
@@ -0,0 +1,156 @@
+  window.addEventListener('message', async function ({data, source}) {
+    if (data.type !== 'archive:req') return
+
+    if (!allow(data.url, data.method)) {
```
Just in case you didn't see it, this is where the permission check is done.

(Although looking at it now, it should be called with `await` to support a user prompt. I'll push a fix for that later.)

pfrazee commented:
Right, I'll probably merge this and solve that later

On Tue, Aug 20, 2019 at 10:39 AM Brecht Savelkoul <notifications@github.com>
wrote:

> *@brechtcs* commented on this pull request.
> ------------------------------
>
> In js/lib/iframe-shim-dat-archive.js
> <https://github.com/beakerbrowser/beaker-sidebar-app/pull/1#discussion_r315762968>
> :
>
> > @@ -0,0 +1,156 @@
> +const STATIC = /^static\:/
> +
> +export function shimDatArchiveGlobal () {
> +  window.DatArchive = DatArchiveShim
> +}
> +
> +export function handleDatArchiveRequests (allow = () => true) {
> +  window.addEventListener('message', async function ({data, source}) {
> +    if (data.type !== 'archive:req') return
> +
> +    if (!allow(data.url, data.method)) {
>
> Just in case you overlooked it, this is where the permission check is done.
>
> (Although looking at it now, it should be called with await to support a
> user prompt. I'll push a fix for that later.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/beakerbrowser/beaker-sidebar-app/pull/1?email_source=notifications&email_token=AAJWCU2CSL3WDGW6BZ2JCDDQFQF4BA5CNFSM4IMG7I3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCDKSVY#pullrequestreview-277260631>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJWCU5M2RAFSSAXUPDEQD3QFQF4BANCNFSM4IMG7I3A>
> .
>

brechtcs commented:
Didn't notice the merge conflicts earlier, should be fixed now.

pfrazee commented:
👍 Will review soon, just wrapping up a bunch of big architectural changes first.
